horace_core/sqw/@sqw/rebin_sqw.m

```diff
@@ -9,438 +9,131 @@
-        elseif ndims==2
-            wout=cut(win,xbin,ybin);
-        elseif ndims==3
-            wout=cut(win,xbin,ybin,[]);
-        elseif ndims==4
-            wout=cut(win,xbin,ybin,[],[]);
-        else
-            error('ERROR: Dimensions of dataset is not integer in the range 1 to 4');
+
+    else
+
+        vec_or_empty = cellfun(@(x) isvector(x) || isempty(x), varargin);
+        if ~all(vec_or_empty)
+            error('HORACE:rebin_sqw:invalid_argument', ...
+                  'check the format of input arguments');
         end
 
-    case 4
         %rebinning x and/or y and/or z axes
-        xbin=varargin{1}; ybin=varargin{2}; zbin=varargin{3};
-        if ~isempty(xbin)
-            if numel(xbin)==1
-                %just a step size specified for x
-            elseif numel(xbin)==3
-                %lo,step,hi
-                if xbin(1)>=xbin(3) || (xbin(3)-xbin(1))<xbin(2)
-                    error('Horace error: problem with specified x-bins. Must be of form [step] or [lo,sttep,hi]');
-                end
-            else
-                xbin=[];
-            end
-        end
-        if ~isempty(ybin)
-            if numel(ybin)==1
-                %just a step size specified for x
-            elseif numel(ybin)==3
-                %lo,step,hi
-                if ybin(1)>=ybin(3) || (ybin(3)-ybin(1))<ybin(2)
-                    error('Horace error: problem with specified y-bins. Must be of form [step] or [lo,sttep,hi]');
-                end
-            else
-                ybin=[];
-            end
-        end
-        if ~isempty(zbin)
-            if numel(zbin)==1
-                %just a step size specified for x
-            elseif numel(zbin)==3
-                %lo,step,hi
-                if zbin(1)>=zbin(3) || (zbin(3)-zbin(1))<zbin(2)
-                    error('Horace error: problem with specified z-bins. Must be of form [step] or [lo,sttep,hi]');
-                end
-            else
-                zbin=[];
-            end
-        end
-        %
-        ndims=dimensions(win);
-        if ndims==1
-            error('Horace error: have specified 2 or more binning arguments for a 1-dimensional object');
-        elseif ndims==2
-            error('Horace error: have specified 3 or more binning arguments for a 2-dimensional object');
-        elseif ndims==3
-            wout=cut(win,xbin,ybin,zbin);
-        elseif ndims==4
-            wout=cut(win,xbin,ybin,zbin,[]);
-        else
-            error('ERROR: Dimensions of dataset is not integer in the range 1 to 4');
+        bins = cellfun(@check_bin, varargin, 'UniformOutput', false);
+
+        ndims = dimensions(win);
+
+        if ndims < numel(bins)
+            error('HORACE:rebin_sqw:invalid_argument', ...
+                  'have specified %d or more binning arguments for a %d-dimensional object', ...
+                  numel(bins), ndims);
         end
 
-end
+    end
+
+    % Pad with empty bins
+    bins = {bins{:} [] [] [] []};
```

Comment:
any particular reason for 4 bins of padding?

Reply:
Technically I only need 3, since 1 arg is always required, but because at maximum I slice 1:4 (for a 4-D rebin, which may only have 1 element in `bins`), I add 4 for safety. This ensures that any under-specification is handled safely.